### ISSUE TITLE
.deb build with cpack doesn't create "hsa" directory in /opt/rocm/include

### ISSUE THREAD

**Issue by c0d3st0rm** [RadeonOpenCompute/ROCR-Runtime#92]
When building this package from source on Ubuntu 20.04 (gcc 9.3.0, cmake 3.16.3, using ROCm tag `roc-3.3.0` for the source clone) after cloning all ROCm related packages using `repo` and the documentation, I can successfully package a `.deb` after following these steps (roughly):
 - `repo init` + `repo sync`
 - build and install `hsakmt-roct` + `hsakmt-roct-dev` to `/opt/rocm`
 - configure an out-of-source ROCR-Runtime build using `cmake -DCPACK_PACKAGING_INSTALL_PREFIX=/opt/rocm -DCMAKE_INSTALL_PREFIX=/opt/rocm -DCMAKE_BUILD_TYPE=Release ../src`
 - `make` (succeeds)
 - `cpack -G DEB` (succeeds)

Trying to install the resulting package, however, fails:
```bash
root@rocm:/rocm/ROCR-Runtime/build# dpkg -i /rocm/ROCR-Runtime/build/hsa-rocr-dev-1.1.9.0-local-build-1a56c09-Linux.deb
Selecting previously unselected package hsa-rocr-dev.
(Reading database ... 21066 files and directories currently installed.)
Preparing to unpack .../hsa-rocr-dev-1.1.9.0-local-build-1a56c09-Linux.deb ...
Unpacking hsa-rocr-dev (1.1.9.0-local-build-1a56c09) ...
dpkg: error processing archive /rocm/ROCR-Runtime/build/hsa-rocr-dev-1.1.9.0-local-build-1a56c09-Linux.deb (--install):
 unable to create '/opt/rocm/include/hsa/Brig.h.dpkg-new' (while processing './opt/rocm/include/hsa/Brig.h'): No such file or directory
Errors were encountered while processing:
 /rocm/ROCR-Runtime/build/hsa-rocr-dev-1.1.9.0-local-build-1a56c09-Linux.deb
```

However, I can manually create the directory at `/opt/rocm/include/hsa`. Doing this means the package can now install successfully:
```bash
root@rocm:/rocm/ROCR-Runtime/build# mkdir /opt/rocm/include/hsa/
root@rocm:/rocm/ROCR-Runtime/build# dpkg -i /rocm/ROCR-Runtime/build/hsa-rocr-dev-1.1.9.0-local-build-1a56c09-Linux.deb
(Reading database ... 21066 files and directories currently installed.)
Preparing to unpack .../hsa-rocr-dev-1.1.9.0-local-build-1a56c09-Linux.deb ...
Unpacking hsa-rocr-dev (1.1.9.0-local-build-1a56c09) ...
Setting up hsa-rocr-dev (1.1.9.0-local-build-1a56c09) ...
Processing triggers for libc-bin (2.31-0ubuntu9) ...
root@rocm:/rocm/ROCR-Runtime/build# 
```

Should this directory be created automatically by the `.deb`, or am I building + installing packages in the wrong order?

**Comment by skeelyamd** [comment]
This is an issue with recent builds of CPACK.  CPACK fails to include the /opt/rocm/include/hsa directory symlink when building DEB packages (RPM packages get this right).  The next release will include a workaround for this issue.

**Comment by c0d3st0rm** [comment]
Would something like this patch applied to the main `CMakeLists.txt` fix the issue? The current Debian `postinst` also hardcodes `/opt/rocm`, which this should fix. I haven't tested the patch though, so I'm not sure if any of the variables, for example, are incorrect.

```
diff --git a/src/CMakeLists.txt b/src/CMakeLists.txt
index 642bd09..ac6c532 100644
--- a/src/CMakeLists.txt
+++ b/src/CMakeLists.txt
@@ -201,10 +201,14 @@ set ( CPACK_PACKAGE_CONTACT "Advanced Micro Devices Inc." )
 set ( CPACK_PACKAGE_DESCRIPTION_SUMMARY "AMD Heterogeneous System Architecture HSA - Linux HSA Runtime for Boltzmann (ROCm) platforms" )
 set ( CPACK_RESOURCE_FILE_LICENSE "${CMAKE_CURRENT_SOURCE_DIR}/LICENSE.md" )
 
+configure_file( DEBIAN/preinst.in DEBIAN/preinst )
+configure_file( DEBIAN/prerm.in DEBIAN/prerm )
+configure_file( DEBIAN/postinst.in DEBIAN/postinst )
+
 # Debian package specific variables
 set ( CPACK_DEBIAN_PACKAGE_DEPENDS "hsakmt-roct" )
 set ( CPACK_DEBIAN_PACKAGE_HOMEPAGE "https://github.com/RadeonOpenCompute/ROCR-Runtime" )
-set ( CPACK_DEBIAN_PACKAGE_CONTROL_EXTRA "${CMAKE_CURRENT_SOURCE_DIR}/DEBIAN/postinst;${CMAKE_CURRENT_SOURCE_DIR}/DEBIAN/prerm" )
+set ( CPACK_DEBIAN_PACKAGE_CONTROL_EXTRA "${CMAKE_CURRENT_BINARY_DIR}/DEBIAN/postinst;${CMAKE_CURRENT_BINARY_DIR}/DEBIAN/prerm;${CMAKE_CURRENT_BINARY_DIR}/DEBIAN/preinst" )
 
 ## RPM package specific variables
 set ( CPACK_RPM_PACKAGE_DEPENDS "hsakmt-roct" )
diff --git a/src/DEBIAN/postinst b/src/DEBIAN/postinst.in
similarity index 68%
rename from src/DEBIAN/postinst
rename to src/DEBIAN/postinst.in
index 769a72e..67f5b86 100644
--- a/src/DEBIAN/postinst
+++ b/src/DEBIAN/postinst.in
@@ -3,7 +3,7 @@
 set -e
 
 do_ldconfig() {
-    echo /opt/rocm/hsa/lib > /etc/ld.so.conf.d/hsa-rocr-dev.conf && ldconfig
+    echo ${CMAKE_INSTALL_PREFIX}/hsa/lib > /etc/ld.so.conf.d/hsa-rocr-dev.conf && ldconfig
 }
 
 case "$1" in
diff --git a/src/DEBIAN/preinst.in b/src/DEBIAN/preinst.in
new file mode 100644
index 0000000..40133da
--- /dev/null
+++ b/src/DEBIAN/preinst.in
@@ -0,0 +1,3 @@
+#!/bin/bash
+
+mkdir -p ${CMAKE_INSTALL_PREFIX}/include/hsa
diff --git a/src/DEBIAN/prerm b/src/DEBIAN/prerm.in
similarity index 100%
rename from src/DEBIAN/prerm
rename to src/DEBIAN/prerm.in
```

**Comment by skeelyamd** [comment]
That should install but will make ${CMAKE_INSTALL_PREFIX}/include/hsa a real directory rather than a symlink.  That may cause build issues with later ROCm components when including ROCr headers.  I'll post the patch we are expecting to correct the issue here.  The reason it hasn't been merged to github yet is that it will conflict with the scripts currently in the binary release packages.

**Comment by skeelyamd** [comment]
https://github.com/RadeonOpenCompute/ROCR-Runtime/commit/ef18428a592ade18b27aa36eb7f065c49b62e03c

**Comment by c0d3st0rm** [comment]
Why does it use symlinks, as opposed to individual directories? What _should_ it be pointing to?

On a related note, when is the next release (3.4.0 I would presume) due?

**Comment by skeelyamd** [comment]
Yes your points-to location is correct.

We use symlinks for historical reasons that many are trying to get rid of.  Unfortunately some components are still using /opt/rocm/hsa/include/hsa so we can't drop that just yet.  We'd like to be using ${CMAKE_INSTALL_PREFIX}/include (or ${CMAKE_INSTALL_PREFIX}/include/hsa, similar to GL & EGL) so that the build can easily target standard install locations and not have /opt/rocm encoded in the build system at all.

**Comment by c0d3st0rm** [comment]
Which packages are still relying on the old paths? Surely that could be cleaned up for _<next release>_? I've noticed some other packages like the profiler (iirc) don't even install into `/opt/rocm`, but `/opt` instead. I've been trying to build ROCm for 20.04, and it's been quite tedious trying to build it all, especially given the experimental build scripts only go up to 2.0.0, and open source packages still depend on closed ones and need the dependencies manually removed (`hsa-ext`) for an open source based release. I will see if I can raise those respective issues up in the proper repositories though.

**Comment by yalue** [comment]
I wanted to voice my support for any such efforts for building all of ROCm from source.  However, I will also add that ROCR-Runtime is one of the easiest packages to build from source, in my opinion!  rocBLAS and MIOpen were, by far, the packages I had the most trouble building (and I think rocBLAS currently won't even compile if you don't have 32 GB of RAM).

A fully open-source ROCm build has been nearly essential for those of us trying to modify the software, since mixing package-manager versions and built-from-source versions makes system updates harder to manage--installing updates risks trashing modified versions of ROCm software, forcing you to rebase any source code modifications and reinstall.  (The new /opt/rocm structure using versioned symlinks should help with this a bit, though.)

**Comment by skeelyamd** [comment]
HIP is going through some very major changes at the moment.  I'm hoping that once this is settled cleaning up the duplicate ROCr paths can be done.  Open sourcing hsa-ext is also in the works.  Once that is done hsa-ext simply won't exist and the functionality will be part of ROCr.  Doing this will add a build dependency on llvm (clang with the amdgpu backend).  We're also looking to improve find_package support.

**Comment by skeelyamd** [comment]
This fix is part of the 3.5.0 update.
